### PR TITLE
Is1273/groups api

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

- new **catalog** service
   - [x] basic
   - [x] CI
   - [x] Add new table of dag to `simcore-postgres-database` package
- **webserver**: new catalog-gateway subsystem
  - [x] api exposed . Increased to 0.3.
  - [x] auth: requires login
  - [x] access: ``services.catalog.*`` permission required
  - [x] routing (reverse-proxy?)

#### Maintenance:
 - web API doc: documented and created tags in openapi 
 - makefile: webserver and catalog service makefile have all ``build`` and ``test`` target flavours
 - adds [shellcheck](https://www.shellcheck.net) tooling in ``scripts/shellcheck`` and vscode configuration in [.vscode-template/settings.json](.vscode-template/settings.json)
 - NEW standard makefile in ``simcore-postgres-database``
       -  highlight: has a setup to create migration commits
  - **IMPORTANT**: this PR requires upgrading deployed databases

## Related issue number
Implements #1273 

## How to test

- catalog service has some unit testing coverage
- catalog subsystem in webserver has some unit-testing converage
```cmd
cd services/catalog
make devel
make install-dev tests
```
- start stacks
```cmd
cd osparc-simcore
make devel
make build up-prod
```
Now open browser

- To see the webserver  OAS: http://127.0.0.1:9081/api/doc
image
- To see the catalog OAS: http://127.0.0.1:8005/docs or http://127.0.0.1:8005/redoc
image
image

## Checklist

- [x] Did you change any service's API? Then make sure to bundle document and upgrade version (``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [x] Unit tests for the changes exist
- [x] Runs in the swarm
- [x] Documentation reflects the changes
- [x] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

File: packages/postgres-database/setup.py [hunk 1]
```diff
@@ -4,58 +4,53 @@
...
+    extras_require={"migration": migration_requirements, "test": test_requirements},
     include_package_data=True,
     package_data={
-        '': [
-            '*.ini',
-            'migration/*.py',
-            'migration/*.mako',
-            'migration/versions/*.py',
-         ]
+        "": ["*.ini", "migration/*.py", "migration/*.mako", "migration/versions/*.py",]
```
Comment (sanderegg) [inline packages/postgres-database/setup.py]: funny this... one line now?

File: scripts/shellcheck [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+#
+# Static analysis tool for scripts
+# - https://www.shellcheck.net
+# - VS extension: https://github.com/timonwong/vscode-shellcheck
+#
+
+exec docker run --rm -i -v "$PWD:/mnt:ro" koalaman/shellcheck:v0.7.0 "$@"
```
Comment (sanderegg) [inline scripts/shellcheck]: very cool, and scarry, too

File: packages/postgres-database/src/simcore_postgres_database/__init__.py [hunk 3]
```diff
@@ -1,8 +1,4 @@
 from . import storage_models, webserver_models
 from .models.base import metadata
 
-__all__ = [
-    'metadata',
-    'webserver_models',
-    'storage_models'
-]
+__all__ = ["metadata", "webserver_models", "storage_models"]
```
Comment (sanderegg) [inline packages/postgres-database/src/simcore_postgres_database/__init__.py]: also oneline here...

File: services/catalog/src/simcore_service_catalog/schemas/project.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import sys
+from enum import Enum
+from pathlib import Path
+from typing import Dict, List, Optional, Union
+
+from pydantic import BaseModel, EmailStr, Field
+
+current_file = Path(sys.argv[0] if __name__ == "__main__" else __file__).resolve()
+
+KEY_RE = r"^(simcore)/(services)(/demodec)?/(comp|dynamic|frontend)(/[^\s]+)+$"
```
Comment (sanderegg) [inline services/catalog/src/simcore_service_catalog/schemas/project.py]: remove demodec
frontend I'm not sure that is still existing?

File: services/catalog/src/simcore_service_catalog/schemas/project.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+VERSION_RE = r"^(0|[1-9]\d*)(\.(0|[1-9]\d*)){2}(-(0|[1-9]\d*|\d*[-a-zA-Z][-\da-zA-Z]*)(\.(0|[1-9]\d*|\d*[-a-zA-Z][-\da-zA-Z]*))*)?(\+[-\da-zA-Z]+(\.[-\da-zA-Z-]+)*)?$"
```
Comment (sanderegg) [inline services/catalog/src/simcore_service_catalog/schemas/project.py]: is that for the version of a node? I don't think  that is compatible with the current way. also it is not in sync with node jsonschema.
I'm also wondering if the integration version should not be in a docker label instead?

File: services/catalog/src/simcore_service_catalog/schemas/project.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+
+class Connection(BaseModel):
+    nodeUuid: Optional[str]
+    output: Optional[str]
+
+
+class FilePickerOutput(BaseModel):
```
Comment (sanderegg) [inline services/catalog/src/simcore_service_catalog/schemas/project.py]: why a FilePicker specific output?

File: services/catalog/src/simcore_service_catalog/schemas/project.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+class Node(BaseModel):
+    key: str = Field(..., regex=KEY_RE, example="simcore/services/comp/sleeper")
+    version: str = Field(..., regex=VERSION_RE, example="6.2.0")
+    label: str = Field(...)
+    progress: float = Field(0, ge=0, le=100)
+    thumbnail: Optional[str]
+
+    inputs: Optional[Dict[str, InputTypes]]
+    inputAccess: Optional[Dict[str, AccessEnum]]
+    inputNodes: List[str] = []
```
Comment (sanderegg) [inline services/catalog/src/simcore_service_catalog/schemas/project.py]: very nice!

File: services/catalog/src/simcore_service_catalog/schemas/schemas_dags.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
+from typing import Dict, Optional
+
+# TODO: why pylint error in pydantic???
```
Comment (sanderegg) [inline services/catalog/src/simcore_service_catalog/schemas/schemas_dags.py]: this TODO? is there any answer?

File: services/catalog/src/simcore_service_catalog/schemas/schemas_dags.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    name: str
+    description: Optional[str]
+    contact: Optional[EmailStr]
+
+
+class DAGIn(DAGBase):
+    workbench: Optional[Dict[str, project.Node]]
+
+
+class DAGInPath(DAGBase):
```
Comment (sanderegg) [inline services/catalog/src/simcore_service_catalog/schemas/schemas_dags.py]: will need some info from you as to what this is...

File: services/catalog/tests/unit/conftest.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+@pytest.fixture()
+def fake_data_dag_in() -> Dict:
+    DAG_DATA_IN_DICT = {
+        "key": "simcore/services/frontend/nodes-group/macros/1",
+        "version": "1.0.0",
+        "name": "string",
+        "description": "string",
+        "contact": "user@example.com",
+        "workbench": {
+            "additionalProp1": {
```
Comment (sanderegg) [inline services/catalog/tests/unit/conftest.py]: additionalProp??

File: services/web/server/src/simcore_service_webserver/config/server-docker-prod.yaml [hunk 11]
```diff
@@ -68,6 +68,10 @@ storage:
   host: ${STORAGE_HOST}
   port: ${STORAGE_PORT}
   version: v0
+catalog:
+  host: catalog
+  port: 8000
+  version: v0
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/config/server-docker-prod.yaml]: no variables?

Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/config/server-docker-prod.yaml]: no. Actually is more informative than actual configurations. E.g. all fastapi based servers are port 8000 and all aiohttp 8080.

File: services/catalog/src/simcore_service_catalog/schemas/schemas_dags.py (hunk 9 above)
Comment (pcrespov) [reply services/catalog/src/simcore_service_catalog/schemas/schemas_dags.py]: This is something amazingly cool :-) we can talk about it on Monday

File: services/catalog/tests/unit/conftest.py (hunk 10 above)
Comment (pcrespov) [reply services/catalog/tests/unit/conftest.py]: It was auto-generated ... Will change when I add more conditions on the keys

File: services/catalog/docker/entrypoint.sh [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+# - Executes *inside* of the container upon start as --user [default root]
+# - Notice that the container *starts* as --user [default root] but
+#   *runs* as non-root user [scu]
+#
+echo $INFO "Entrypoint for stage ${SC_BUILD_TARGET} ..."
+echo "  User    :`id $(whoami)`"
+echo "  Workdir :`pwd`"
+
+
+if [[ ${SC_BUILD_TARGET} == "development" ]]
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/catalog/docker/entrypoint.sh]: this is old .. now all shells have the benediction of shellcheck! :-)

File: packages/postgres-database/src/simcore_postgres_database/__init__.py (hunk 3 above)
Comment (pcrespov) [reply packages/postgres-database/src/simcore_postgres_database/__init__.py]: black measures horizontal and vertical spaces and decides whether it is inline or multiline

File: services/catalog/src/simcore_service_catalog/schemas/project.py (hunk 4 above)
Comment (pcrespov) [reply services/catalog/src/simcore_service_catalog/schemas/project.py]: For the moment i keep compatibility with projects.yaml ...

File: services/catalog/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+ENV SC_BUILD_TARGET production
+ENV SC_BOOT_MODE production
+
+WORKDIR /home/scu
+
+RUN mkdir -p services/catalog &&\
+    chown scu:scu services/catalog &&\
+    mv /build/services/catalog/docker services/catalog/docker &&\
+    rm -rf /build
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/catalog/Dockerfile]: Agreed, but the final image will not have that folder anymore ... which could be confusing. I think it does not harm, oder?

**Pull request merged**